### ISSUE TITLE
Memory problems

### ISSUE THREAD

**Issue by Vlader97** [bergerhealer/TrainCarts#220]
##### Spigot version: 1.12.2-R0.1

<hr>

#### Problem or bug:
When using the automatic train control system, running trains allocate more and more memory and the garbage collector does not help. The result is that my server crash because all free memory was used. In 1 day, it also takes 1GB of RAM. I tested it on an empty server.
#### Expected behaviour:
I expect trains to get the memory needed for traffic, but not all the time!
#### Steps to reproduce:
I have 1 km long metro system under my city and the trains have 5 carts and on line was 4 trains.
Trains have the following properties: Speedlimit 1, keepchunks true, playercollision cancel other setting is default. I have 10 security devices on the track like [waiter] which is set to 100 blocks and every sign is always powered in my metro. And in plugin setting is set keepChunksLoadedOnlyWhenMoving: false and other setting is without change.

**Comment by bbayu123** [comment]
1. How much memory did you allocate?
2. Can you provide detailed timings on the server?

**Comment by bbayu123** [comment]
From the details you have given me so far, there are a number of possible things that may be causing lag:
1. UpdateDirection is called too frequently due to train speed being too high.
2. BlockCollision (a parameter added in the development builds) being DEFAULT is introducing a lot of lag due to train hitbox intersecting with nearby blocks (is only true if you use models).
3. Long distance waiters make searching for trains ahead consume server tick routines.

**Comment by Vlader97** [comment]
1. The server is set on 4GB.
2. I don't know if I understand what you want. If is server empty, trains allocate memory at a rate of several MB per hour. In 1 day after start server using 1GB more that it was on start.
3.  I don't have problems with lags.

**Comment by bbayu123** [comment]
As I said, I need the timings of the server when the memory is high. Do [/timings on] to start timings, and [/timings paste] to get the timings.

**Comment by bbayu123** [comment]
Anyways, the only thing I'm concerned about is TrainCarts using only 1 GB of RAM is far too less due to algorithm logic.

**Comment by Vlader97** [comment]
Oh I did not know about this command. I will send timing later.
Did you say it's 1GB a little? 1GB of RAM per day is really a lot. How much do you think is enough? I have no unlimited memory.
He also said that plugin caches a lot of things into memory banks.  And do you delete data from those banks when the plugin don't needed? I doubt that plugin need more data on the same track.

**Comment by bbayu123** [comment]
I believe memory is allocated per train, not per track.

**Comment by Vlader97** [comment]
That's possible, I do not know how it's implemented in the plugin.
If in those memory banks accumulating unnecessary data, then their cleaning will solve this problem.

**Comment by bbayu123** [comment]
I don't think anything that the trains cache are unnecessary... although I can't say for sure, developer is offline at the moment.

**Comment by bergerkiller** [comment]
Theres several caching performed to speed up performance overall:
- Rail type and positions detected in the world
- Signs detected for rails
- Rails blocks in use for each minecart
All this data is routinely wiped and as a life of about 10 ticks unless actually used. So while a train is moving, it will continuously poll the same rails, keeping that data loaded. When the train is destroyed, and no other train is driving over the same rails within 10 ticks, the data cached there is gone.

Memory leak from the plugin is unlikely. However, a memory leak due to an increasing number of chunks being loaded is a possibility. Can you run some diagnostics tools to see how many chunks are loaded over time? I don't know what tools can do so as I'm out of the loop, all I know about is NoLagg which is largely no longer maintained.

**Comment by Vlader97** [comment]
Using the /lag command in the essentials plugin shows that the number of active bits is between 350-400. But the memory grows, I do not know what tool to measure it.
There are some timings:
https://www.spigotmc.org/go/timings?url=mobajiqecu
https://timings.spigotmc.org/?url=mobajiqecu
https://timings.spigotmc.org/?url=yusimilixi

**Comment by bergerkiller** [comment]
You can tell java to generate a memory dump when it runs out of memory. Add the following java commandline parameters to that for your server:
`-XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=coredump.dmp`

Let it run for a while until it runs out of memory, then it will generate a 1GB file in the location where Spigot (server) jar is located. Zip that file and upload it somewhere.

If you cant/dont want to send it, you can analyze it yourself using the [jhat binary](https://docs.oracle.com/javase/7/docs/technotes/tools/share/jhat.html) that comes with the Java 8 installation. It requires a deep understanding of Java to interpret it though.

**Comment by Vlader97** [comment]
Ok the file has over 3 GB (compressed for transfer to 600MB). You can download here: https://www.dropbox.com/s/ol9br8iakpzyvxy/TRAIN.tar.gz?dl=0

**Comment by Vlader97** [comment]
Has the file helped to find the problem?
I think you were right about those chunks when I don't use keepchunks so it does not happen.

**Comment by bbayu123** [comment]
I mean, if you place a redstone lamp at the location of the sign, whether the lamp is powered will show whether the sign receives power and thus whether it will activate or not.

**Comment by bbayu123** [comment]
Lol wrong issue

**Comment by bergerkiller** [comment]
I've had to fix many things since then so I haven't had real time to properly inspect it (I did download your coredump, so you don't have to keep it available in dropbox).

If keep chunks loaded = true exacerbates the problem, then the likely problem is that chunks don't unload properly when trains go by them. I wonder if one of those 'lag fixer' plugins that routinely unload chunks without players nearby can help you here. It won't harm the keep-chunks-loaded logic, since TC will cancel the chunk unload when a train is nearby them anyway.
